core/icons/mutator_icon.ts

```diff
@@ -0,0 +1,358 @@
+   *     I.E. the block that owns this icon.
+   */
+  private getBubbleOwnerRect(): Rect {
+    const bbox = this.sourceBlock.getSvgRoot().getBBox();
+    return new Rect(bbox.y, bbox.y + bbox.height, bbox.x, bbox.x + bbox.width);
+  }
+
+  /** Decomposes the source block to create blocks in the mini workspace. */
+  private createRootBlock() {
+    this.rootBlock = this.sourceBlock.decompose!(
+      this.miniWorkspaceBubble!.getWorkspace()
+    )!;
+
+    for (const child of this.rootBlock.getDescendants(false)) {
+      child.queueRender();
+    }
+
+    this.rootBlock.setMovable(false);
+    this.rootBlock.setDeletable(false);
+
+    const flyoutWidth =
+      this.miniWorkspaceBubble?.getWorkspace()?.getFlyout()?.getWidth() ?? 0;
+    this.rootBlock.moveBy(
+      this.rootBlock.RTL ? -(flyoutWidth + this.MARGIN) : this.MARGIN,
+      this.MARGIN
+    );
+  }
+
+  /** Adds a listen to the source block that triggers saving connections. */
+  private addSaveConnectionsListener() {
+    if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+    const saveConnectionsListener = () => {
+      if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+      this.sourceBlock.saveConnections(this.rootBlock);
+    };
+    saveConnectionsListener();
+    this.sourceBlock.workspace.addChangeListener(saveConnectionsListener);
+  }
+
+  /**
+   * Creates a change listener to add to the mini workspace which recomposes
+   * the block.
+   */
+  private createMiniWorkspaceChangeListener() {
+    return (e: Abstract) => {
+      if (!this.shouldIgnoreMutatorEvent(e) && !this.updateWorkspacePid) {
+        this.updateWorkspacePid = setTimeout(() => {
+          this.updateWorkspacePid = null;
+          this.recomposeSourceBlock();
+        }, 0);
+      }
+    };
+  }
+
+  /**
+   * Returns true if the given event is not one the mutator needs to
+   * care about.
+   */
+  private shouldIgnoreMutatorEvent(e: Abstract) {
+    return (
+      e.isUiEvent ||
+      e.type === eventUtils.CREATE ||
+      (e.type === eventUtils.CHANGE &&
+        (e as BlockChange).element === 'disabled')
+    );
+  }
+
+  /** Recomposes the source block based on changes to the mini workspace. */
+  private recomposeSourceBlock() {
+    if (!this.rootBlock) return;
+
+    const existingGroup = eventUtils.getGroup();
+    if (!existingGroup) eventUtils.setGroup(true);
+
+    const oldExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+    this.sourceBlock.compose!(this.rootBlock);
+    const newExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+
+    if (oldExtraState !== newExtraState) {
+      eventUtils.fire(
+        new (eventUtils.get(eventUtils.BLOCK_CHANGE))(
+          this.sourceBlock,
+          'mutation',
+          null,
+          oldExtraState,
+          newExtraState
+        )
+      );
+    }
+
+    eventUtils.setGroup(existingGroup);
+  }
+
+  /** @internal */
```

Comment:
After updating the tests and seeing where this is accessed, I'm wondering if it actually should be provided publicly. I just hate to provide it when I know all of the functionality is not supported.

Reply:
I don't know what to think about this.  I can see arguments in both directions.  But now I'm wondering why `.miniWorkspaceBubble` is `private`.  Making it `protected` would aid subclassing—but I'd guessed it was public if I'd not gone and checked just now.

Reply:
Haha turns out that `getWorkspace` is currently internal, so no reason to start providing it unless we get an ask.